File: 1319-number-of-operations-to-make-network-connected/1319-number-of-operations-to-make-network-connected.cpp
class DisjointSet{
public:
    vector<int> parent, size;

    DisjointSet(int n){
        size.resize(n, 1);
        parent.resize(n);

        for(int i = 0; i < n; i++){
            parent[i] = i;
        }
    }

    int find(int u){
        if(parent[u] == u) return u;

        return parent[u] = find(parent[u]);
    }

    bool unionUV(int u, int v){
        int rootU = find(u);
        int rootV = find(v);

        if(rootU == rootV) return false;

        if(size[rootU] < size[rootV]){
            parent[rootU] = rootV;
            size[rootV] += size[rootU];
        } else {
            parent[rootV] = rootU;
            size[rootU] += size[rootV];
        }

        return true;
    }
};

class Solution {
public:
    int makeConnected(int n, vector<vector<int>>& connections) {
        DisjointSet s1(n);

        int numDisconnectedComponents = 0;
        int extraEdges = 0;

        for(const auto &connection : connections){
            int u = connection[0];
            int v = connection[1];

            if(!s1.unionUV(u,v)) extraEdges++;
        }

        for(int i = 0; i < n; i++){
            if(s1.parent[i] == i) numDisconnectedComponents++;
        }

        return (extraEdges >= numDisconnectedComponents - 1) ? numDisconnectedComponents - 1 : -1;
        
    }
};

File: 1319-number-of-operations-to-make-network-connected/1319-number-of-operations-to-make-network-connected.py
class DisjointSet:
    def __init__(self, n):
        self.parent = list(range(n+1))
        self.size = [1]*(n+1)

    def find(self, u):
        if(self.parent[u] == u):
            return u
        
        self.parent[u] = self.find(self.parent[u])
        return self.parent[u]

    def union(self, u, v):
        rootU = self.find(u)
        rootV = self.find(v)

        if(rootU == rootV):
            return False

        if(self.size[rootU] < self.size[rootV]):
            self.parent[rootU] = rootV
            self.size[rootV] += self.size[rootU]
        else:
            self.parent[rootV] = rootU
            self.size[rootU] += self.size[rootV]
        
        return True



class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        s1 = DisjointSet(n)

        extraEdges = 0
        connectedComponents = 0

        for u,v in connections:
            if not s1.union(u,v):
                extraEdges += 1

        for i in range(n):
            if s1.find(i) == i:
                connectedComponents += 1
        

        return connectedComponents -1 if (extraEdges >= connectedComponents -1) else -1